libbpf-tools/cpufreq.c

```diff
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Wenbo Zhang
+//
+// Based on cpufreq(8) from BPF-Perf-Tools-Book by Brendan Gregg.
+// 10-OCT-2020   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <linux/perf_event.h>
+#include <asm/unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "cpufreq.h"
+#include "cpufreq.skel.h"
+#include "trace_helpers.h"
+
+#define FREQ	99
+
+static struct env {
+	int duration;
+	bool verbose;
+} env = {
+	.duration = -1
+};
+
+const char *argp_program_version = "cpufreq 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+const char argp_program_doc[] =
+"Sampling CPU freq system-wide & by process. Ctrl-C to end.\n"
+"\n"
+"USAGE: cpufreq [--help] [-d DURATION]\n"
+"\n";
+
+static const struct argp_option opts[] = {
+	{ "duration", 'd', "DURATION", 0, "Duration to trace"},
```

Comment:
please specify units, is it seconds? Always confusing to guess whether it is milliseconds, seconds, something else?

Reply:
Fixed, thanks! I'll check others tools to make sure no confusing.